### PR TITLE
Scale tank mix component rates based on total rate 

### PR DESCRIPTION
**Pull request by freejelliott**
The example code provided by Climate suggests scaling the individually defined product rates by the overall rate of the whole tank mix. This implements that plus rewrites some other parts of code to be similar to the example.

### CODE REVIEW

**Comment by freejelliott** [comment]
@Qingnan-Toby-Liu I don't expect a functional review of this, but if you could point out if anything is blatantly wrong C#/.NET-wise that would be great :)

File: Services/ParserService.cs [hunk 1]
```diff
@@ -477,70 +481,56 @@ public static List<AsHarvestedSummary> GetAsHarvestedDataSummaries(ApplicationDa
...
-                if (product.ProductComponents.Count > 1)
+                if (product.ProductComponents.Count > 0)
                 {
                     System.Console.WriteLine("Product Components:");
 
                     UnitSystem.UnitOfMeasureConverter converter = new UnitSystem.UnitOfMeasureConverter();
 
                     // Use the numerator from the rate - either gal or lbs to match the component units
                     UnitSystem.UnitOfMeasure targetUnits = appUom.Components[0].Unit;
+                    if (appUom.Components[1].Unit.Label != "ac") {
```
Comment (Qingnan-Toby-Liu) [inline Services/ParserService.cs:492]: Are you sure appUom.Components count is >= 1? You may want to use safe navigation 
if (appUom.Components.Count() > 0 && !String.IsNullOrEmpty(appUom.Components[1]?.Unit) &&  appUom.Components[1].Unit.Label != "ac") {

File: Services/ParserService.cs [hunk 2]
```diff
@@ -477,70 +481,56 @@ public static List<AsHarvestedSummary> GetAsHarvestedDataSummaries(ApplicationDa
...
-                        if (componentUnits.UnitDimension == targetUnits.UnitDimension)
+                        // If they are different dimensions (i.e. mass and volume) then we can't do anything to scale
+                        // the component rates. Just take them as they are.
+                        if (componentUnits.UnitDimension != targetUnits.UnitDimension)
                         {
-                            // Normalize the component units to be the same as the application units
-                            double normalizedRate = converter.Convert(componentUnits, targetUnits, component.Quantity.Value.Value);
-                            totalRate += normalizedRate;
+                            totalRate = averageRate;
+                            System.Console.WriteLine("expected dimension " + targetUnits.UnitDimension.Name + " " + targetUnits.Label + " got " + componentUnits.UnitDimension.Name+ " " + componentUnits.Label);
```
Comment (Qingnan-Toby-Liu) [inline Services/ParserService.cs:510]: You may want to try to use String interpolation for better performance if there are a lot of ProductComponent:
string str = $"expected dimension {targetUnits.UnitDimension.Name} {targetUnits.Label} got {componentUnits.UnitDimension.Name} {componentUnits.Label}";

**Comment by jdelard** [comment]
image

File: Services/ParserService.cs [hunk 3]
```diff
@@ -382,26 +382,42 @@ public static List<AsHarvestedSummary> GetAsHarvestedDataSummaries(ApplicationDa
...
+                            System.Console.WriteLine("name: " + meter.Representation.Code);
+                            System.Console.WriteLine("desc: " + meter.Representation.Description);
+                            System.Console.WriteLine("long desc: " + meter.Representation.LongDescription);
+                            System.Console.WriteLine("unit code: " + appRateMeter.UnitOfMeasure.Code);
+                            System.Console.WriteLine("unit dim: " + appRateMeter.UnitOfMeasure.Dimension.ToString());
+                            System.Console.WriteLine("unit offset: " + appRateMeter.UnitOfMeasure.Offset.ToString());
+                            System.Console.WriteLine("unit scale: " + appRateMeter.UnitOfMeasure.Scale.ToString());
+                            System.Console.WriteLine();
+                            break;
                     }
```
[the remark this answers is not in the document]
Comment (freejelliott) [reply Services/ParserService.cs:420]: I believe in this case we have a list like:
```
a, b, c, d, e, f
```
and we want to assign variables to `b` and `e`.
So we iterate over the list and check if the current item matches what we're looking for. If it doesn't that's fine, we just move onto the next item.

I will add a check that both the variables are not null when the loop finishes.

File: Services/ParserService.cs [hunk 4]
```diff
@@ -382,26 +382,42 @@ public static List<AsHarvestedSummary> GetAsHarvestedDataSummaries(ApplicationDa
                 System.Console.WriteLine("Pass Name: " + opData.Description);
             }
 
-            Dictionary<string, WorkingData> codeToMeter = new Dictionary<string, WorkingData>();
-            IEnumerable<DeviceElementUse> headDEUs = opData.GetDeviceElementUses(2);
+            WorkingData areaMeter = null;
```
[the remark this answers is not in the document]
Comment (freejelliott) [reply Services/ParserService.cs:385]: As per the above comment I'll do a check after the loop that it's not null

File: Services/ParserService.cs [hunk 5]
```diff
@@ -411,54 +427,42 @@ public static List<AsHarvestedSummary> GetAsHarvestedDataSummaries(ApplicationDa
...
-
-                    if (kv.Key == "vrDeltaArea")
-                    {
-                        area = nrv.Value.Value;
-                    } else if (kv.Key == "vrAppRateVolumeActual" || kv.Key == "vrAppRateMassActual") 
-                    {
-                        rate = nrv.Value.Value;
-                    }
-                }
+                NumericRepresentationValue area = spatialRecord.GetMeterValue(areaMeter) as NumericRepresentationValue;
```
[the remark this answers is not in the document]
Comment (freejelliott) [reply Services/ParserService.cs:444]: Probably not, but no longer a problem after the check

File: Services/ParserService.cs (hunk 1 above)
Comment (freejelliott) [reply Services/ParserService.cs:492]: I'll add a check earlier on to assert this

**Review by Qingnan-Toby-Liu — APPROVED** [review]

**Pull request merged**